mmedit/models/editors/disco_diffusion/guider.py

```diff
@@ -0,0 +1,484 @@
+                clamp_grad=True,
+                clamp_max=0.05,
+                clip_guidance_scale=5000):
+        """Clip guidance function.
+
+        Args:
+            model (nn.Module): _description_
+            diffuser (object): _description_
+            x (torch.Tensor): _description_
+            t (int): _description_
+            beta_prod_t (torch.Tensor): _description_
+            model_stats (List[torch.Tensor]): _description_
+            secondary_model (nn.Module): A smaller secondary diffusion model
+                trained by Katherine Crowson to remove noise from intermediate
+                timesteps to prepare them for CLIP.
+                Ref: https://twitter.com/rivershavewings/status/1462859669454536711 # noqa
+                Defaults to None.
+            init_image (torch.Tensor): Initial image for denoising.
+                Defaults to None.
+            clamp_grad (bool, optional): Whether clamp gradient. Defaults to True.
+            clamp_max (float, optional): Clamp max values. Defaults to 0.05.
+            clip_guidance_scale (int, optional): The scale of influence of
+                clip guidance on image generation. Defaults to 5000.
+        """
+        with torch.enable_grad():
+            x_is_NaN = False
+            x = x.detach().requires_grad_()
+            n = x.shape[0]
+            if secondary_model is not None:
+                alpha = torch.tensor(
+                    diffuser.alphas_cumprod[t]**0.5, dtype=torch.float32)
+                sigma = torch.tensor(
+                    (1 - diffuser.alphas_cumprod[t])**0.5, dtype=torch.float32)
+                cosine_t = alpha_sigma_to_t(alpha, sigma).to(x.device)
+                model_output = secondary_model(
+                    x, cosine_t[None].repeat([x.shape[0]]))
+                pred_original_sample = model_output['pred']
+            else:
+                model_output = model(x, t)['outputs']
+                model_output, predicted_variance = torch.split(
+                    model_output, x.shape[1], dim=1)
+                alpha_prod_t = 1 - beta_prod_t
+                pred_original_sample = (x - beta_prod_t**(0.5) *
+                                        model_output) / alpha_prod_t**(0.5)
+            # fac = diffuser_output['beta_prod_t']** (0.5)
+            # x_in = diffuser_output['original_sample'] * fac + x * (1 - fac)
+            fac = beta_prod_t**(0.5)
+            x_in = pred_original_sample * fac + x * (1 - fac)
+            x_in_grad = torch.zeros_like(x_in)
+            for model_stat in model_stats:
+                for i in range(self.cutter_cfg.cutn_batches):
+                    t_int = int(t.item()) + 1
+                    try:
+                        input_resolution = model_stat[
+                            'clip_model'].model.visual.input_resolution
+                    except AttributeError:
+                        input_resolution = 224
+
+                    cuts = MakeCutoutsDango(
```

Comment:
Cutters are rebuilt for all timesteps, why not make this a function?

Reply:
Or we can make `MakeCutoutsDango.forward` takes `timestep` as input and calculate the number of `Overview`, `innercrop`, `IC_size_Pow`, and `IC_Grey_P` should be used.

Reply:
True, now I move cutter settings to `infer` arguments.